### PR TITLE
crack Villager RNG.

### PR DESCRIPTION
**Pull request by reply7251**
using DynamicProgram so need to change build.gradle.

usage:
Villager plays Amethyst Chime sound when moving on Amethyst blocks.

so let a villager stand on 3 Amethyst blocks in a row,
build double piston to make villager move on Amethyst blocks,
and the timer for piston should be more than 20ticks(10gt) to make sure there's Amethyst Chime sound every push.
(if the server is lag you should increase the interval)

need a 20hz clock for sync with server.
then use `/ccrackvillager clock "clock position"` to crack RNG
use `/ccrackvillager addgoal` to add your desire trade result (and price)
and `/ccrackvillager run` to crack once

video:
[video](https://youtu.be/tljuQ7VHAqo)

### CODE REVIEW

File: src/main/java/net/earthcomputer/clientcommands/command/CrackVillagerRNGCommand.java [hunk 1]
```diff
@@ -2,42 +2,85 @@
...
+        return Command.SINGLE_SUCCESS;
+    }
+
+    private static int clearGoals(CommandContext<FabricClientCommandSource> context) {
+        CCrackVillager.goalOffers.clear();
+        return Command.SINGLE_SUCCESS;
+    }
+
+    private static int removeGoal(CommandContext<FabricClientCommandSource> context) {
+        context.getSource().sendFeedback(Component.literal("remove goal: " + CCrackVillager.goalOffers.remove(getInteger(context, "index")-1)));
```
Comment (RealRTTV) [inline src/main/java/net/earthcomputer/clientcommands/command/CrackVillagerRNGCommand.java]: please use a translatable text

Comment (RealRTTV) [inline src/main/java/net/earthcomputer/clientcommands/command/CrackVillagerRNGCommand.java]: reformat to add spacing

```suggestion
        context.getSource().sendFeedback(Component.literal("remove goal: " + CCrackVillager.goalOffers.remove(getInteger(context, "index") - 1)));
```

File: src/main/java/net/earthcomputer/clientcommands/command/CrackVillagerRNGCommand.java [hunk 2]
```diff
@@ -2,42 +2,85 @@
...
+        return Command.SINGLE_SUCCESS;
     }
 
-    private static int lookingForEnchantment(FabricClientCommandSource source, ItemAndEnchantmentsPredicate predicate) {
-        CCrackVillager.targetEnchantment = predicate;
+    private static int listGoals(CommandContext<FabricClientCommandSource> context) {
+        for(var i = 0; i < CCrackVillager.goalOffers.size(); i++) {
+            var offer = CCrackVillager.goalOffers.get(i);
+            context.getSource().sendFeedback(Component.literal("%d. %s".formatted(i+1, offer)));
```
Comment (RealRTTV) [inline src/main/java/net/earthcomputer/clientcommands/command/CrackVillagerRNGCommand.java]:
```suggestion
            context.getSource().sendFeedback(Component.literal("%d: %s".formatted(i + 1, offer)));
```
also make translate

File: src/main/java/net/earthcomputer/clientcommands/command/CrackVillagerRNGCommand.java [hunk 3]
```diff
@@ -62,4 +105,97 @@ private static int setInterval(FabricClientCommandSource source, int interval) t
         CCrackVillager.setInterval(interval);
         return Command.SINGLE_SUCCESS;
     }
+
+    static class MyItemArgument extends ItemArgument {
```
Comment (RealRTTV) [inline src/main/java/net/earthcomputer/clientcommands/command/CrackVillagerRNGCommand.java]: please make separate class, or use existing `ItemAndEnchantmentsArgument.class` along with `CItemStackPredicateArgument.class`

File: src/main/java/net/earthcomputer/clientcommands/command/CrackVillagerRNGCommand.java [hunk 4]
```diff
@@ -2,42 +2,85 @@
...
-                        .executes(ctx -> lookingForEnchantment(ctx.getSource(), getItemAndEnchantmentsPredicate(ctx, "name"))))));
+            .then(literal("cancel")
+                .executes(ctx -> cancel(ctx.getSource())))
+            .then(literal("clock")
+                .then(argument("clockpos", blockPos())
+                    .executes(ctx -> crackVillagerRNG(ctx.getSource(), getBlockPos(ctx, "clockpos")))))
+            .then(literal("interval")
+                .then(argument("ticks", IntegerArgumentType.integer(0, 20))
+                    .executes(ctx -> setInterval(ctx.getSource(), getInteger(ctx, "ticks")))))
+            .then(literal("addgoal")
```
Comment (RealRTTV) [inline src/main/java/net/earthcomputer/clientcommands/command/CrackVillagerRNGCommand.java]:
```suggestion
            .then(literal("add-goal")
```

File: src/main/java/net/earthcomputer/clientcommands/command/CrackVillagerRNGCommand.java [hunk 5]
```diff
@@ -2,42 +2,85 @@
...
+                .then(genFirst(context))
+                .then(genSecond(context))
+                .then(genResult(context)))
+            .then(literal("listgoals")
```
Comment (RealRTTV) [inline src/main/java/net/earthcomputer/clientcommands/command/CrackVillagerRNGCommand.java]:
```suggestion
            .then(literal("list-goals")
```

File: src/main/java/net/earthcomputer/clientcommands/command/CrackVillagerRNGCommand.java [hunk 6]
```diff
@@ -2,42 +2,85 @@
...
+                .executes(CrackVillagerRNGCommand::listGoals))
+            .then(literal("removegoal")
```
Comment (RealRTTV) [inline src/main/java/net/earthcomputer/clientcommands/command/CrackVillagerRNGCommand.java]:
```suggestion
            .then(literal("remove-goal")
```

File: src/main/java/net/earthcomputer/clientcommands/command/CrackVillagerRNGCommand.java [hunk 7]
```diff
@@ -2,42 +2,85 @@
...
+                .then(argument("index", integer(1,CCrackVillager.goalOffers::size))
+                    .executes(CrackVillagerRNGCommand::removeGoal)))
+            .then(literal("cleargoals").executes(CrackVillagerRNGCommand::clearGoals))
```
Comment (RealRTTV) [inline src/main/java/net/earthcomputer/clientcommands/command/CrackVillagerRNGCommand.java]:
```suggestion
            .then(literal("clear-goals").executes(CrackVillagerRNGCommand::clearGoals))
```

File: src/main/java/net/earthcomputer/clientcommands/command/CrackVillagerRNGCommand.java [hunk 8]
```diff
@@ -62,4 +105,97 @@ private static int setInterval(FabricClientCommandSource source, int interval) t
...
+    }
+
+
+
+    static LiteralArgumentBuilder<FabricClientCommandSource> genFirst(CommandBuildContext context) {
+        var item = item(context);
+        Supplier<Integer> supplier = () -> item.lastItem.getItem().getDefaultMaxStackSize();
+        return literal("first")
+            .then(argument("firstitem", withString(combined(item, combined(integer(1,supplier).allowAny(), integer(1,supplier).allowAny()))))
+                .executes(CrackVillagerRNGCommand::setGoal)
```
Comment (RealRTTV) [inline src/main/java/net/earthcomputer/clientcommands/command/CrackVillagerRNGCommand.java:154]: wait, so a valid goal can be a first item with no specific result or anything else set?

File: src/main/java/net/earthcomputer/clientcommands/command/arguments/CombinedArgument.java [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+import com.mojang.brigadier.context.CommandContext;
+import com.mojang.brigadier.exceptions.CommandSyntaxException;
+import com.mojang.brigadier.exceptions.SimpleCommandExceptionType;
+import com.mojang.brigadier.suggestion.Suggestions;
+import com.mojang.brigadier.suggestion.SuggestionsBuilder;
+import net.minecraft.network.chat.Component;
+
+import java.util.concurrent.CompletableFuture;
+
+public class CombinedArgument<A, B> implements ArgumentType<CombinedArgument.Combined<A, B>> {
```
Comment (RealRTTV) [inline src/main/java/net/earthcomputer/clientcommands/command/arguments/CombinedArgument.java:14]: why is this required?

File: src/main/java/net/earthcomputer/clientcommands/command/arguments/DynamicIntegerArgument.java [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+import com.mojang.brigadier.arguments.ArgumentType;
+import com.mojang.brigadier.context.CommandContext;
+import com.mojang.brigadier.exceptions.CommandSyntaxException;
+import com.mojang.brigadier.suggestion.Suggestions;
+import com.mojang.brigadier.suggestion.SuggestionsBuilder;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Supplier;
+
+public class DynamicIntegerArgument implements ArgumentType<Integer> {
```
Comment (RealRTTV) [inline src/main/java/net/earthcomputer/clientcommands/command/arguments/DynamicIntegerArgument.java:13]: you don't need this just use CRangeArgument and predicate if the range is valid from the stack's max size, also if you were to need it, use `MinMaxBounds.Ints`

File: src/main/java/net/earthcomputer/clientcommands/command/arguments/DynamicIntegerArgument.java [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+    }
+
+    public static DynamicIntegerArgument integer(Supplier<Integer> limit) {
+        return integer(0, limit);
+    }
+
+    @Override
+    public Integer parse(final StringReader reader) throws CommandSyntaxException {
+        final int start = reader.getCursor();
+        if(allowAny && reader.canRead() && reader.peek() == '*') {
```
Comment (RealRTTV) [inline src/main/java/net/earthcomputer/clientcommands/command/arguments/DynamicIntegerArgument.java]:
```suggestion
        if (allowAny && reader.canRead() && reader.peek() == '*') {
```

File: src/main/java/net/earthcomputer/clientcommands/command/arguments/DynamicIntegerArgument.java [hunk 12]
```diff
@@ -0,0 +1,77 @@
...
+        if (result > max) {
+            reader.setCursor(start);
+            throw CommandSyntaxException.BUILT_IN_EXCEPTIONS.integerTooHigh().createWithContext(reader, result, max);
+        }
+        return result;
+    }
+
+    @Override
+    public <S> CompletableFuture<Suggestions> listSuggestions(CommandContext<S> context, SuggestionsBuilder builder) {
+        if(allowAny) {
```
Comment (RealRTTV) [inline src/main/java/net/earthcomputer/clientcommands/command/arguments/DynamicIntegerArgument.java]:
```suggestion
        if (allowAny) {
```

File: src/main/java/net/earthcomputer/clientcommands/command/CrackVillagerRNGCommand.java [hunk 13]
```diff
@@ -62,4 +105,97 @@ private static int setInterval(FabricClientCommandSource source, int interval) t
...
+                .then(genSecond(context)) .then(genResult(context)));
+    }
+
+    static  LiteralArgumentBuilder<FabricClientCommandSource> genSecond(CommandBuildContext context) {
+        var item = item(context);
+        Supplier<Integer> supplier = () -> item.lastItem.getItem().getDefaultMaxStackSize();
+        return literal("second")
+            .then(argument("seconditem", withString(combined(item, combined(integer(1,supplier).allowAny(), integer(1,supplier).allowAny()))))
```
Comment (RealRTTV) [inline src/main/java/net/earthcomputer/clientcommands/command/CrackVillagerRNGCommand.java]:
```suggestion
            .then(argument("seconditem", withString(combined(item, combined(integer(1, supplier).allowAny(), integer(1, supplier).allowAny()))))
```

File: src/main/java/net/earthcomputer/clientcommands/command/arguments/EnchantmentArgument.java [hunk 14]
```diff
@@ -0,0 +1,39 @@
...
+import com.mojang.brigadier.suggestion.Suggestions;
+import com.mojang.brigadier.suggestion.SuggestionsBuilder;
+import net.minecraft.commands.SharedSuggestionProvider;
+import net.minecraft.core.registries.BuiltInRegistries;
+import net.minecraft.resources.ResourceLocation;
+import net.minecraft.world.item.enchantment.Enchantment;
+
+import java.util.concurrent.CompletableFuture;
+
+public class EnchantmentArgument implements ArgumentType<Enchantment> {
```
Comment (RealRTTV) [inline src/main/java/net/earthcomputer/clientcommands/command/arguments/EnchantmentArgument.java:16]: `ItemAndEnchantmentsPredicateArgument` left the chat

File: src/main/java/net/earthcomputer/clientcommands/features/CCrackVillager.java [hunk 15]
```diff
@@ -94,10 +96,16 @@ static void crack(ClientboundSoundPacket packet) {
                 }
                 onCrackFinished.accept(VillagerRNGSim.INSTANCE.getSeed());
             } else {
-                Minecraft.getInstance().gui.setOverlayMessage(Component.translatable("commands.ccrackvillager.fail"), false);
+                var gui = Minecraft.getInstance().gui;
+                if(clockPos == null) {
```
Comment (RealRTTV) [inline src/main/java/net/earthcomputer/clientcommands/features/CCrackVillager.java:104]:
```suggestion
                if (clockPos == null) {
```

File: src/main/java/net/earthcomputer/clientcommands/features/CCrackVillager.java [hunk 16]
```diff
@@ -130,4 +138,64 @@ static void reset() {
         measurements.clear();
         validMeasures = 0;
     }
+
+    public static class Offer implements Predicate<MerchantOffer> {
+
```
Comment (RealRTTV) [inline src/main/java/net/earthcomputer/clientcommands/features/CCrackVillager.java]: why the empty line?

File: src/main/java/net/earthcomputer/clientcommands/features/CCrackVillager.java [hunk 17]
```diff
@@ -130,4 +138,64 @@ static void reset() {
...
+        public Offer withSecond(Predicate<ItemStack> predicate, String description) {
+            second = predicate;
+            secondDescription = description;
+            return this;
+        }
+        public Offer withResult(Predicate<ItemStack> predicate, String description) {
+            result = predicate;
+            resultDescription = description;
+            return this;
+        }
```
Comment (RealRTTV) [inline src/main/java/net/earthcomputer/clientcommands/features/CCrackVillager.java:172]:
```suggestion
        }

```

File: src/main/java/net/earthcomputer/clientcommands/features/CCrackVillager.java [hunk 18]
```diff
@@ -130,4 +138,64 @@ static void reset() {
...
+            first = predicate;
+            firstDescription = description;
+            return this;
+        }
+        public Offer withSecond(Predicate<ItemStack> predicate, String description) {
+            second = predicate;
+            secondDescription = description;
+            return this;
+        }
+        public Offer withResult(Predicate<ItemStack> predicate, String description) {
```
Comment (RealRTTV) [inline src/main/java/net/earthcomputer/clientcommands/features/CCrackVillager.java:168]:
```suggestion

        public Offer withResult(Predicate<ItemStack> predicate, String description) {
```

File: src/main/java/net/earthcomputer/clientcommands/features/CCrackVillager.java [hunk 19]
```diff
@@ -130,4 +138,64 @@ static void reset() {
...
+        String firstDescription = "";
+        String secondDescription = "";
+        String resultDescription = "";
+        String enchantmentDescription = "";
+
+        public Offer withFirst(Predicate<ItemStack> predicate, String description) {
+            first = predicate;
+            firstDescription = description;
+            return this;
+        }
```
Comment (RealRTTV) [inline src/main/java/net/earthcomputer/clientcommands/features/CCrackVillager.java:160]:
```suggestion
        }

```

File: src/main/java/net/earthcomputer/clientcommands/features/CCrackVillager.java [hunk 20]
```diff
@@ -130,4 +138,64 @@ static void reset() {
...
+        public Offer andEnchantment(Predicate<ItemStack> predicate, String description) {
+            result = result.and(predicate);
+            enchantmentDescription = description;
+            return this;
+        }
+
+        @Override
+        public boolean test(MerchantOffer offer) {
+            if(first != null && !first.test(offer.getItemCostA().itemStack())) return false;
+            if(second != null && !second.test(offer.getItemCostB().isPresent() ? offer.getItemCostB().get().itemStack() : null)) {
```
Comment (RealRTTV) [inline src/main/java/net/earthcomputer/clientcommands/features/CCrackVillager.java]:
```suggestion
            if (second != null && !second.test(offer.getItemCostB().isPresent() ? offer.getItemCostB().get().itemStack() : null)) {
```

Comment (RealRTTV) [reply src/main/java/net/earthcomputer/clientcommands/features/CCrackVillager.java]: a tenary does not belong here, please

File: src/main/java/net/earthcomputer/clientcommands/features/CCrackVillager.java [hunk 21]
```diff
@@ -130,4 +138,64 @@ static void reset() {
...
+        }
+        public Offer andEnchantment(Predicate<ItemStack> predicate, String description) {
+            result = result.and(predicate);
+            enchantmentDescription = description;
+            return this;
+        }
+
+        @Override
+        public boolean test(MerchantOffer offer) {
+            if(first != null && !first.test(offer.getItemCostA().itemStack())) return false;
```
Comment (RealRTTV) [inline src/main/java/net/earthcomputer/clientcommands/features/CCrackVillager.java]:
```suggestion
            if (first != null && !first.test(offer.getItemCostA().itemStack())) return false;
```

File: src/main/java/net/earthcomputer/clientcommands/features/CCrackVillager.java [hunk 22]
```diff
@@ -130,4 +138,64 @@ static void reset() {
...
+            if(second != null && !second.test(offer.getItemCostB().isPresent() ? offer.getItemCostB().get().itemStack() : null)) {
+                return false;
+            }
+            return result == null || result.test(offer.getResult());
+        }
+
+        @Override
+        public String toString() {
+            StringBuilder result = new StringBuilder(firstDescription);
+            if(!secondDescription.isEmpty() && !result.isEmpty()) {
```
Comment (RealRTTV) [inline src/main/java/net/earthcomputer/clientcommands/features/CCrackVillager.java]:
```suggestion
            if (!secondDescription.isEmpty() && !result.isEmpty()) {
```

File: src/main/java/net/earthcomputer/clientcommands/features/CCrackVillager.java [hunk 23]
```diff
@@ -130,4 +138,64 @@ static void reset() {
...
+                result.append(" + ");
+            }
+            result.append(secondDescription);
+            if(!resultDescription.isEmpty() && !result.isEmpty()) {
```
Comment (RealRTTV) [inline src/main/java/net/earthcomputer/clientcommands/features/CCrackVillager.java]:
```suggestion
            if (!resultDescription.isEmpty() && !result.isEmpty()) {
```

File: src/main/java/net/earthcomputer/clientcommands/features/VillagerRNGSim.java [hunk 24]
```diff
@@ -34,6 +34,9 @@ public class VillagerRNGSim {
     int ticksToWait = 0;
     boolean synced = false;
 
+    static long lastBruteForce = 0;
+    static long bruteForceResult = 0;
```
Comment (RealRTTV) [inline src/main/java/net/earthcomputer/clientcommands/features/VillagerRNGSim.java]: why package-private?

File: src/main/java/net/earthcomputer/clientcommands/features/VillagerRNGSim.java [hunk 25]
```diff
@@ -62,8 +65,16 @@ public void onAmethyst(ClientboundSoundPacket packet) {
 
             errorCount = 0;
             if(ticks == 0) {
```
Comment (RealRTTV) [inline src/main/java/net/earthcomputer/clientcommands/features/VillagerRNGSim.java]:
```suggestion
            if (ticks == 0) {
```

File: src/main/java/net/earthcomputer/clientcommands/features/VillagerRNGSim.java [hunk 26]
```diff
@@ -62,8 +65,16 @@ public void onAmethyst(ClientboundSoundPacket packet) {
...
+                if((tickCounter - lastBruteForce > 500 || tickCounter > bruteForceResult + 10) && CCrackVillager.findingOffers) {
```
Comment (RealRTTV) [inline src/main/java/net/earthcomputer/clientcommands/features/VillagerRNGSim.java]:
```suggestion
                if ((tickCounter - lastBruteForce > 500 || tickCounter > bruteForceResult + 10) && CCrackVillager.findingOffers) {
```

File: src/main/java/net/earthcomputer/clientcommands/features/VillagerRNGSim.java [hunk 27]
```diff
@@ -62,8 +65,16 @@ public void onAmethyst(ClientboundSoundPacket packet) {
...
+                    lastBruteForce = tickCounter;
+                    bruteForce();
+                }
                 synced = true;
-                Minecraft.getInstance().gui.setOverlayMessage(Component.translatable("commands.ccrackvillager.synced"), false);
+                if(CCrackVillager.findingOffers) {
```
Comment (RealRTTV) [inline src/main/java/net/earthcomputer/clientcommands/features/VillagerRNGSim.java]:
```suggestion
                if (CCrackVillager.findingOffers) {
```

File: src/main/java/net/earthcomputer/clientcommands/features/VillagerRNGSim.java [hunk 28]
```diff
@@ -145,7 +157,7 @@ public void onTick(boolean sim) {
         }
         nextInt(100);
 
-        if(CCrackVillager.cracked && !sim && CCrackVillager.targetEnchantment != null && synced) {
+        if(CCrackVillager.cracked && !sim && CCrackVillager.findingOffers && synced) {
```
Comment (RealRTTV) [inline src/main/java/net/earthcomputer/clientcommands/features/VillagerRNGSim.java]:
```suggestion
        if (CCrackVillager.cracked && !sim && CCrackVillager.findingOffers && synced) {
```

File: src/main/java/net/earthcomputer/clientcommands/features/VillagerRNGSim.java [hunk 29]
```diff
@@ -145,7 +157,7 @@ public void onTick(boolean sim) {
...
             var player = Minecraft.getInstance().player;
             var villager = CCrackVillager.targetVillager.get();
             if(player == null || player.distanceTo(villager) > 5) return;
```
Comment (RealRTTV) [inline src/main/java/net/earthcomputer/clientcommands/features/VillagerRNGSim.java]:
```suggestion
            if (player == null || player.distanceTo(villager) > 5) {
                return;
            }
```

File: src/main/java/net/earthcomputer/clientcommands/features/CCrackVillager.java (hunk 21 above)
Comment (RealRTTV) [inline src/main/java/net/earthcomputer/clientcommands/features/CCrackVillager.java]:
```suggestion
            if (first != null && !first.test(offer.getItemCostA().itemStack())) {
                return false;
            }
```

File: src/main/java/net/earthcomputer/clientcommands/features/VillagerRNGSim.java [hunk 30]
```diff
@@ -156,19 +168,45 @@ public void onTick(boolean sim) {
             var offers = simulate.predictOffers();
             if(offers == null) return;
             for(var offer : offers) {
-                if(CCrackVillager.targetEnchantment.test(offer.getResult())) {
+                if(CCrackVillager.goalOffers.stream().anyMatch(goalOffer -> goalOffer.test(offer))) {
```
Comment (RealRTTV) [inline src/main/java/net/earthcomputer/clientcommands/features/VillagerRNGSim.java]:
```suggestion
                if (CCrackVillager.goalOffers.stream().anyMatch(goalOffer -> goalOffer.test(offer))) {
```

File: src/main/java/net/earthcomputer/clientcommands/features/VillagerRNGSim.java [hunk 31]
```diff
@@ -156,19 +168,45 @@ public void onTick(boolean sim) {
             var offers = simulate.predictOffers();
             if(offers == null) return;
             for(var offer : offers) {
```
Comment (RealRTTV) [inline src/main/java/net/earthcomputer/clientcommands/features/VillagerRNGSim.java]:
```suggestion
            for (var offer : offers) {
```

File: src/main/java/net/earthcomputer/clientcommands/features/VillagerRNGSim.java [hunk 32]
```diff
@@ -156,19 +168,45 @@ public void onTick(boolean sim) {
             var offers = simulate.predictOffers();
             if(offers == null) return;
```
Comment (RealRTTV) [inline src/main/java/net/earthcomputer/clientcommands/features/VillagerRNGSim.java]:
```suggestion
            if (offers == null) {
                return;
            }
```

File: src/main/java/net/earthcomputer/clientcommands/features/VillagerRNGSim.java [hunk 33]
```diff
@@ -156,19 +168,45 @@ public void onTick(boolean sim) {
...
+    void bruteForce() {
+        var player = Minecraft.getInstance().player;
+        var villager = CCrackVillager.targetVillager.get();
+        if(player == null || player.distanceTo(villager) > 5) return;
+        var simulate = clone();
+
+        for(var i = 0; i < CCrackVillager.interval; i++) {
+            simulate.onTick(true);
+        }
+        for(var step = 0; step < 600; step++) {
```
Comment (RealRTTV) [inline src/main/java/net/earthcomputer/clientcommands/features/VillagerRNGSim.java]:
```suggestion
        for (int step = 0; step < 600; step++) {
```

File: src/main/java/net/earthcomputer/clientcommands/features/VillagerRNGSim.java [hunk 34]
```diff
@@ -156,19 +168,45 @@ public void onTick(boolean sim) {
...
                     break;
                 }
             }
         }
     }
 
+    void bruteForce() {
+        var player = Minecraft.getInstance().player;
+        var villager = CCrackVillager.targetVillager.get();
+        if(player == null || player.distanceTo(villager) > 5) return;
```
Comment (RealRTTV) [inline src/main/java/net/earthcomputer/clientcommands/features/VillagerRNGSim.java]:
```suggestion
        if (player == null || player.distanceTo(villager) > 5) {
            return;
        }
```

File: src/main/java/net/earthcomputer/clientcommands/features/VillagerRNGSim.java [hunk 35]
```diff
@@ -156,19 +168,45 @@ public void onTick(boolean sim) {
...
-                    chat.addMessage(Component.literal("I found it !"));
-                    CCrackVillager.targetEnchantment = null;
+                    CCrackVillager.findingOffers = false;
                     break;
                 }
             }
         }
     }
 
+    void bruteForce() {
```
Comment (RealRTTV) [inline src/main/java/net/earthcomputer/clientcommands/features/VillagerRNGSim.java:184]: - no var please
- why package-private?

File: src/main/java/net/earthcomputer/clientcommands/features/VillagerRNGSim.java [hunk 36]
```diff
@@ -156,19 +168,45 @@ public void onTick(boolean sim) {
...
+        for(var i = 0; i < CCrackVillager.interval; i++) {
+            simulate.onTick(true);
+        }
+        for(var step = 0; step < 600; step++) {
+            simulate.onTick(true);
+            var sim2 = simulate.clone();
+
+            var offers = sim2.predictOffers();
+            if(offers == null) return;
+            for(var offer : offers) {
```
Comment (RealRTTV) [inline src/main/java/net/earthcomputer/clientcommands/features/VillagerRNGSim.java]:
```suggestion
            for (var offer : offers) {
```

File: src/main/java/net/earthcomputer/clientcommands/features/VillagerRNGSim.java [hunk 37]
```diff
@@ -156,19 +168,45 @@ public void onTick(boolean sim) {
...
+                for(var goal : CCrackVillager.goalOffers) {
```
Comment (RealRTTV) [inline src/main/java/net/earthcomputer/clientcommands/features/VillagerRNGSim.java]:
```suggestion
                for (var goal : CCrackVillager.goalOffers) {
```

File: src/main/java/net/earthcomputer/clientcommands/features/VillagerRNGSim.java [hunk 38]
```diff
@@ -156,19 +168,45 @@ public void onTick(boolean sim) {
...
+                    if(goal.test(offer)) {
```
Comment (RealRTTV) [inline src/main/java/net/earthcomputer/clientcommands/features/VillagerRNGSim.java]:
```suggestion
                    if (goal.test(offer)) {
```

File: src/main/java/net/earthcomputer/clientcommands/features/VillagerRNGSim.java [hunk 39]
```diff
@@ -156,19 +168,45 @@ public void onTick(boolean sim) {
...
+                        bruteForceResult = sim2.tickCounter;
+                        Minecraft.getInstance().gui.getChat().addMessage(Component.translatable("commands.ccrackvillager.match", step));
```
Comment (RealRTTV) [inline src/main/java/net/earthcomputer/clientcommands/features/VillagerRNGSim.java]: please consider using sendFeedback

File: src/main/java/net/earthcomputer/clientcommands/features/VillagerRNGSim.java [hunk 40]
```diff
@@ -156,19 +168,45 @@ public void onTick(boolean sim) {
...
+                        return;
+                    }
+                }
+            }
+        }
+        Minecraft.getInstance().gui.getChat().addMessage(Component.translatable("commands.ccrackvillager.noMatch"));
```
Comment (RealRTTV) [inline src/main/java/net/earthcomputer/clientcommands/features/VillagerRNGSim.java]: same here

File: src/main/resources/assets/clientcommands/lang/en_us.json [hunk 41]
```diff
@@ -267,10 +267,16 @@
   "commands.ccrackvillager.success": "Villager RNG cracked: %d",
   "commands.ccrackvillager.maintain": "Villager RNG maintain success",
   "commands.ccrackvillager.synced": "Villager RNG synced",
+  "commands.ccrackvillager.syncedNotCracking": "Villager RNG synced (not cracking trades)",
   "commands.ccrackvillager.cancel": "Crack RNG cancelled",
   "commands.ccrackvillager.maintainFail": "Villager RNG maintain fail",
   "commands.ccrackvillager.syncWithLag": "set delay to %d tick(s)",
   "commands.ccrackvillager.showEnchOnTick": "tick: %d => %s",
+  "commands.ccrackvillager.noMatch": "cannot find any trades match any goals in 30 seconds",
```
Comment (RealRTTV) [inline src/main/resources/assets/clientcommands/lang/en_us.json]: sort alphabetically please

**Review by RealRTTV — CHANGES_REQUESTED** [review]
very quick scan, needs work on styling

File: src/main/java/net/earthcomputer/clientcommands/command/arguments/EnchantmentArgument.java (hunk 14 above)
Comment (reply7251) [reply src/main/java/net/earthcomputer/clientcommands/command/arguments/EnchantmentArgument.java:16]: I remember that `ItemAndEnchantmentsPredicateArgument` doesn't have enchantment suggestions, and doesn't support enchanted book?

Comment (RealRTTV) [reply src/main/java/net/earthcomputer/clientcommands/command/arguments/EnchantmentArgument.java:16]: no, and no.

**Comment by Earthcomputer** [comment]
Superseded by #642
